Content.Server/JoinQueue/JoinQueueManager.cs

```diff
@@ -0,0 +1,125 @@
+
+namespace Content.Server.JoinQueue;
+
+public sealed class JoinQueueManager
+{
+    [Dependency] private readonly IPlayerManager _playerManager = default!;
+    [Dependency] private readonly IConnectionManager _connectionManager = default!;
+    [Dependency] private readonly IConfigurationManager _cfg = default!;
+    [Dependency] private readonly IServerNetManager _netManager = default!;
+    
+    /// <summary>
+    ///     Queue of active player sessions
+    /// </summary>
+    private readonly List<IPlayerSession> _queue = new();
+    
+    private bool _isEnabled = false;
+    
+    public int PlayerInQueueCount => _queue.Count;
+    public int ActualPlayersCount => _playerManager.PlayerCount - PlayerInQueueCount; // Now it's only real value with actual players count that in game
+
+    public void Initialize()
+    {
+        _netManager.RegisterNetMessage<MsgQueueUpdate>();
+        
+        _cfg.OnValueChanged(CCVars.QueueEnabled, OnQueueCVarChanged, true);
+        _playerManager.PlayerStatusChanged += OnPlayerStatusChanged;
+    }
+    
+    private void OnQueueCVarChanged(bool value)
+    {
+        _isEnabled = value;
+
+        if (!value)
+        {
+            foreach (var session in _queue)
+            {
+                session.ConnectedClient.Disconnect("Queue was disabled");
+            }
+        }
+    }
+
+    private async void OnPlayerStatusChanged(object? sender, SessionStatusEventArgs e)
+    {
+        if (e.NewStatus == SessionStatus.Connected)
+        {
+            if (!_isEnabled)
+            {
+                SendToGame(e.Session);
+                return;
+            }
+                
+            var isPrivileged = await _connectionManager.HavePrivilegedJoin(e.Session.UserId);
+            var haveFreeSlot = _playerManager.PlayerCount < _cfg.GetCVar(CCVars.SoftMaxPlayers);
+            if (isPrivileged || haveFreeSlot)
+            {
+                SendToGame(e.Session);
+                return;
+            }
+                
+            _queue.Add(e.Session);
+            ProcessQueue(false);
+        }
+
+        if (e.NewStatus == SessionStatus.Disconnected)
+        {
+            _queue.Remove(e.Session);
+            ProcessQueue(true);
+        }
+    }
+    
+    /// <summary>
+    ///     If possible, takes the first player in the queue and sends him into the game
+    /// </summary>
+    private void ProcessQueue(bool isDisconnect)
+    {
+        var players = ActualPlayersCount;
+        if (isDisconnect)
+            players--; // Decrease currently disconnected session but that has not yet been deleted
+        
+        var haveFreeSlot = players < _cfg.GetCVar(CCVars.SoftMaxPlayers);
+        var queueContains = _queue.Count > 0;
+        if ((!_isEnabled || haveFreeSlot) && queueContains)
+        {
+            var session = _queue.First();
+            _queue.Remove(session);
+            SendToGame(session);
+        }
+
+        SendUpdateMessages();
+    }
```

Comment:
Is this better off just using an actual Queue<T>?

Reply:
Actual Queue can't `.Remove` disconnected players